lms/lmsweb/views.py

```diff
@@ -83,27 +92,88 @@ def ratelimit_handler(e):
     f'{LIMITS_PER_MINUTE}/minute;{LIMITS_PER_HOUR}/hour',
     deduct_when=lambda response: response.status_code != 200,
 )
-def login(login_error: Optional[str] = None):
-
+def login(login_message: Optional[str] = None):
     if current_user.is_authenticated:
         return get_next_url(request.args.get('next'))
 
     username = request.form.get('username')
     password = request.form.get('password')
     next_page = request.form.get('next')
-    login_error = request.args.get('login_error')
-    user = User.get_or_none(username=username)
+    login_message = request.args.get('login_message')
 
     if request.method == 'POST':
-        if user is not None and user.is_password_valid(password):
+        try:
+            user = auth(username, password)
+        except (ForbiddenPermission, UnauthorizedError) as e:
+            error_message, _ = e.args
+            error_details = {'next': next_page, 'login_message': error_message}
+            return redirect(url_for('login', **error_details))
+        else:
             login_user(user)
             return get_next_url(next_page)
-        elif user is None or not user.is_password_valid(password):
-            login_error = 'Invalid username or password'
-            error_details = {'next': next_page, 'login_error': login_error}
-            return redirect(url_for('login', **error_details))
 
-    return render_template('login.html', login_error=login_error)
+    return render_template('login.html', login_message=login_message)
+
+
+@webapp.route('/signup', methods=['GET', 'POST'])
+def signup():
+    if not webapp.config.get('REGISTRATION_OPEN', False):
+        return redirect(url_for(
+            'login', login_message=_('לא ניתן להירשם כעת'),
+        ))
+
+    form = RegisterForm()
+    if not form.validate_on_submit():
+        return render_template('signup.html', form=form)
+
+    user = User.create(**{
+        User.mail_address.name: form.email.data,
+        User.username.name: form.username.data,
+        User.fullname.name: form.fullname.data,
+        User.role.name: Role.get_unverified_role(),
+        User.password.name: form.password.data,
+        User.api_key.name: User.random_password(),
+    })
+    send_confirmation_mail(user)
+    return redirect(url_for(
+        'login', login_message=_('ההרשמה בוצעה בהצלחה'),
+    ))
+
+
+@webapp.route('/confirm-email/<int:user_id>/<token>')
+def confirm_email(user_id: int, token: str):
+    user = User.get_or_none(User.id == user_id)
+    if user is None:
```

Comment:
The right thing to do it probably to return a generic message which says "The authentication code is invalid"